### PR TITLE
Smartico Bid Adapter: add new bid adapter

### PR DESCRIPTION
**Pull request by cs83**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by ChrisHuie** [comment]
@cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/2812

**Comment by cs83** [comment]
Hello, 
Could you please tell if there is something left we have to commit in order our adapter to become live?

**Comment by cs83** [comment]
> @cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html

@bretg  Hello, could you tell if we need to add something else in order our code to be merged?

**Comment by cs83** [comment]
> > @cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html
> 
> @bretg Hello, could you tell if we need to add something else in order our code to be merged?

@bretg Hello, we have added documnetation to https://github.com/prebid/prebid.github.io and made a pull request https://github.com/prebid/prebid.github.io/pull/2812, could you please check it, if it is the that you expected.

**Comment by cs83** [comment]
> > > @cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html
> > 
> > 
> > @bretg Hello, could you tell if we need to add something else in order our code to be merged?
> 
> @bretg Hello, we have added documnetation to https://github.com/prebid/prebid.github.io and made a pull request [prebid/prebid.github.io#2812](https://github.com/prebid/prebid.github.io/pull/2812), could you please check it, if it is the that you expected.



> > @cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html
> 
> @bretg Hello, could you tell if we need to add something else in order our code to be merged?

@bretg: this is the mentioned documetation pull request link https://github.com/prebid/prebid.github.io/pull/2812,

**Comment by cs83** [comment]
> > @cs83 can you please also submit a documentation pr for this new adapter to our docs repo https://github.com/prebid/prebid.github.io. Here are additional instructions at the bottom of the page if you need them https://docs.prebid.org/dev-docs/bidder-adaptor.html
> 
> @bretg Hello, could you tell if we need to add something else in order our code to be merged?

@bretg Hello, we have added documnetation to https://github.com/prebid/prebid.github.io and made a pull request prebid/prebid.github.io#2812, could you please check it, if it is the that you expected.

**Review by osazos — COMMENTED** [review]
Hi, 

In addition to my comments, can you increase your test coverage, Prebid.js guidelines require at least 80%.

When I tested your adapter with the basic `/gpt/hello_world.html` page I get a CORS error on your endpoint `https://trmads.eu/preBidRequest`. Can you check it?

File: modules/smarticoBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  method: 'POST'
+}
+
+const BIDDER_CODE = 'smartico';
+
+export const spec = {
+  code: BIDDER_CODE,
+  supportedMediaTypes: [BANNER],
+  isBidRequestValid: function (bid) {
+    return !!(bid.params.token && bid.params.placementId);
```
[the remark this answers is not in the document]
Comment (cs83) [reply modules/smarticoBidAdapter.js]: fixed

File: modules/smarticoBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          delete bidParam.regions
+        }
+        bidParams.push(bidParam)
+      }
+    }
+
+    var ServerRequestObjects = {
+      method: SMARTICO_CONFIG.method,
+      url: SMARTICO_CONFIG.bidRequestUrl,
+      bids: validBidRequests,
```
[the remark this answers is not in the document]
Comment (cs83) [reply modules/smarticoBidAdapter.js:68]: 'bids' are added in order to be available in the 'interpretResponse' function (as property of the second argument), no need to send them.

File: modules/smarticoBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      data: {bidParams: bidParams, auctionId: bidderRequest.auctionId}
+    }
+
+    return ServerRequestObjects;
+  },
+  interpretResponse: function (serverResponse, bidRequest) {
+    function getBidById(responceBidId) {
```
[the remark this answers is not in the document]
Comment (cs83) [reply modules/smarticoBidAdapter.js]: fixed

File: modules/smarticoBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    var html
+    var ad
+    var token
+    var language
+    var scriptId
+    var bidRespones = []
+
+    for (i = 0; i < serverResponse.length; i++) {
+      ad = serverResponse[i];
+      bid = getBidById(ad.bidId);
```
[the remark this answers is not in the document]
Comment (cs83) [reply modules/smarticoBidAdapter.js]: refactored as suggested

File: modules/smarticoBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+          requestId: bid.bidId,
+          cpm: ad.cpm,
+          width: parseInt(ad.bannerFormatWidth),
+          height: parseInt(ad.bannerFormatHeight),
+          creativeId: ad.id,
+          netRevenue: false, // gross
+          ttl: ad.ttl,
+          ad: html
+        }
+        bidRespones.push(bidObject);
```
[the remark this answers is not in the document]
Comment (cs83) [reply modules/smarticoBidAdapter.js]: fixed

**Comment by cs83** [comment]
> Hi,
> 
> In addition to my comments below, can you increase your test coverage, Prebid.js guidelines require at least 80%.
> 
> When I tested your adapter with the basic `/gpt/hello_world.html` page I get a CORS error on your endpoint `https://trmads.eu/preBidRequest`. Can you check it?

Hi,  updated the endpoint, the CORS error should be no more.

**Comment by osazos** [comment]
Thanks,

I still get CORS error, 

smartico-cors

After that, just a few line to lint and we should be good :)

```
/home/circleci/Prebid.js/modules/smarticoBidAdapter.js
  58:1   error  Expected indentation of 12 spaces but found 13  indent
  74:60  error  Block must not be padded by blank lines         padded-blocks
  89:1   error  Trailing spaces not allowed                     no-trailing-spaces
  91:1   error  Trailing spaces not allowed                     no-trailing-spaces

```

**Comment by cs83** [comment]
> Thanks,
> 
> I still get CORS error,
> 
> smartico-cors
> 
> After that, just a few line to lint and we should be good :)
> 
> ```
> /home/circleci/Prebid.js/modules/smarticoBidAdapter.js
>   58:1   error  Expected indentation of 12 spaces but found 13  indent
>   74:60  error  Block must not be padded by blank lines         padded-blocks
>   89:1   error  Trailing spaces not allowed                     no-trailing-spaces
>   91:1   error  Trailing spaces not allowed                     no-trailing-spaces
> ```
Hello, i have made changes accordingly. please check it again.

**Comment by osazos** [comment]
I am sorry, I still got a CORS error, see below.

Do you have this issue when you try your bidder with `./integrationExamples/gpt/hello_world.html` ?

smartico-cors-bis

**Review by osazos — APPROVED** [review]
LGTM

**Comment by osazos** [comment]
I can't fetch your last commit from your fork. Please ensure the modification is applied.

**Comment by cs83** [comment]
> @cs83 I can't fetch your last commit from your fork. Please ensure the modification is applied.

Hello, Added one more commit. please  try to fetch.

**Comment by cs83** [comment]
> @cs83 I can't fetch your last commit from your fork. Please ensure the modification is applied.

@osazos Hell, did you manage to fetch changes?

**Comment by cs83** [comment]
@bretg, @ChrisHuie Hello, could you tell if we need to do something more as seems that all checks are passed?

**Review by osazos — APPROVED** [review]

**Comment by osazos** [comment]
@cs83 Sorry for the delay. 
There is no more CORS issue, so it is ok for me.
I'll merge it.

**Pull request merged**